Add SQL Server support

fredwang1012 opened the pull request:
- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB], [MINORBUMP]
- [ ] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, visit our [community resources](https://docs.greatexpectations.io/docs/core/introduction/community_resources#contribute-code-or-documentation).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

stejin committed:
https://support.greatexpectations.io/hc/en-us/requests/210

stejin committed:
Merge branch 'great-expectations:develop' into stejin_01

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stejin committed:
Merge

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stejin committed:
Updated test_connection

stejin committed:
Merge branch 'develop' into stejin_01

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

stejin committed:
Merge branch 'stejin_01' of github.com:stejin/great_expectations into stejin_01

fredwang1012 committed:
Fixed SQL Server SQL expectation 'batch' error

fredwang1012 committed:
Enabled regex expectations for SQL Server

fredwang1012 committed:
SQL Server view compatibility fix

fredwang1012 committed:
Restored functionality that was lost from last commit

fredwang1012 committed:
Restored some accidentally removed functionality in previous commits

fredwang1012 committed:
Fixed an issue with {batch} problem

fredwang1012 committed:
Some further SQL Server compatibility improvements

fredwang1012 committed:
Schema now wrapped in [brackets] in SQL query

fredwang1012 committed:
Fixed some comments

fredwang1012 committed:
Removed some unnecessary bits

stejin committed:
Merge branch 'great-expectations:develop' into stejin_01

stejin committed:
Merge branch 'great-expectations:develop' into stejin_01

fredwang1012 committed:
Refactored SQL Server Regex

fredwang1012 committed:
Merge branch 'stejin_01' into fredwang1012

stejin committed:
Merge pull request #4 from fredwang1012/fredwang1012

Fredwang1012

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

fredwang1012 committed:
Regex fix

fredwang1012 committed:
Fixed regex

stejin committed:
Merge pull request #5 from fredwang1012/fredwang1012

Cleaned up regex fix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

fredwang1012 committed:
Merge remote-tracking branch 'great-expectations/develop' into fredwang1012

fredwang1012 committed:
Merge remote-tracking branch 'great-expectations/develop' into fredwang1012

fredwang1012 committed:
Enhanced regex support

fredwang1012 committed:
Comprehensive implementation of regex to like

fredwang1012 committed:
Fixed batch problem with Databricks

fredwang1012 committed:
Finalized regex to like function

fredwang1012 committed:
Merge remote-tracking branch 'great-expectations/develop' into fredwang1012

fredwang1012 committed:
Merge upstream/develop into fredwang1012

fredwang1012 committed:
Minor clean ups

fredwang1012 commented:
@cla-bot check

fredwang1012 commented:
@cla-bot check

fredwang1012 commented:
@cla-bot check

tyler-hoffman commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -124,6 +127,15 @@ def to_lower_if_not_quoted(
             return value
+
+    # Check MSSQL brackets
+    if value.startswith("[") and value.endswith("]"):
```
We should have a mechanism to support this in mssql-specific code, rather than in our general sql code, since this is only applicable to ms sql. We'll likely want to introduce a new top level datasource type, like we've done for other SQL datasources.

That said, I think we can merge your PR with this, and we will add an internal ticket to move to subclassing.

tyler-hoffman commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -998,51 +1010,109 @@ class TableAsset(_SQLAsset):
         description="Name of the SQL table. Will default to the value of `name` if not provided.",
     )
-    schema_name: Optional[str] = None
+    schema_name: Optional[Any] = None  # Any because validator may transform to quoted_name
```
It looks like `quoted_name` extends `str`, so we shouldn't need to loosen typing on this. Alternatively, we can use `Union[str, quoted_name]`, but AFAIK that is unnecessary and could run into issues with our compatibility layer, given that sqlalchemy won't be installed in all environments.

tyler-hoffman commented on great_expectations/execution_engine/sqlalchemy_batch_data.py:
```diff
@@ -244,22 +244,10 @@ def _create_temporary_table(  # noqa: C901, PLR0912 # FIXME CoP
-            stmt = f"{querymod[0]}into {{temp_table_name}} from{querymod[1]}".format(
-                temp_table_name=temp_table_name
-            )
+            stmt = f"SELECT * INTO {temp_table_name} FROM ({query}) AS subquery"
```
This looks like a behavioral change. I think we should keep the previous behavior of splitting on the FROM for cases where the query doesn't select by `*`.

tyler-hoffman commented on great_expectations/execution_engine/sqlalchemy_execution_engine.py:
```diff
@@ -328,6 +328,28 @@ def __init__(  # noqa: C901, PLR0912, PLR0913, PLR0915 # FIXME CoP
             )
         elif self.dialect_name == GXSqlDialect.DATABRICKS:
             self.dialect_module = import_library_module("databricks.sqlalchemy")
+            # Patch Databricks dialect to use backticks instead of double quotes
```
I'm surprised to see DataBricks updates in this PR. Is this a separate bug? Can you move this to another PR?

tyler-hoffman commented on great_expectations/expectations/metrics/column_map_metrics/column_values_match_regex_list.py:
```diff
@@ -46,11 +46,12 @@ def _pandas(cls, column, regex_list, match_on, **kwargs):
     @column_condition_partial(engine=SqlAlchemyExecutionEngine)
     def _sqlalchemy(cls, column, regex_list, match_on, _dialect, **kwargs):
         if match_on not in ["any", "all"]:
-            raise ValueError("match_on must be any or all")  # noqa: TRY003 # FIXME CoP
```
Non-blocking, but it'd be good to keep some of these white-spaces out of this PR to keep the number of modified files down.

tyler-hoffman commented on great_expectations/expectations/metrics/util.py:
```diff
@@ -109,16 +109,1188 @@ def _is_databricks_dialect(dialect: ModuleType | sa.Dialect | Type[sa.Dialect])
         pass
     return False
 
+def regex_to_like(regex):
```
I'm concerned about the maintenance burden with this. Let's at least move this into its own PR and add a ton of unit tests, along with some high level integration tests.

tyler-hoffman commented on great_expectations/expectations/regex_based_column_map_expectation.py:
```diff
@@ -84,13 +85,19 @@ def _pandas(cls, column, **kwargs):
-            logger.warning(msg)
-            raise NotImplementedError(msg)
-
+            logger.warning(f"Regex is not supported for dialect {_dialect.dialect.name!s}")
```
Any reason not to keep this info in the exception?

tyler-hoffman commented on great_expectations/expectations/metrics/util.py:
```diff
@@ -475,15 +1668,35 @@ def column_reflection_fallback(  # noqa: C901, PLR0912, PLR0915 # FIXME CoP
             ).alias("sys_tables_table_clause")
+
+            # views query
+            views_table_clause: sqlalchemy.TableClause = sa.table(
```
Does SQLAlchemy not support this type of metadata fetching for MSSQL? We should at least get some tests that hit this part of the code.

tyler-hoffman commented:
Hey @fredwang1012 ! Thanks for the contribution! I have a couple requests
* Can we break out the DataBricks updates to a separate PR?
* Similar with the regex -> like functionality
* We're striving for stronger typing in the codebase, so if possible, I'd like to see those `Any`s moved back to `str`. I think that should work, given that the quoted names type in sqlalchemy looks to subclass `str`
* Can you add some testing around this functionality? https://github.com/great-expectations/great_expectations/tree/develop/tests/integration/data_sources_and_expectations/data_sources contains integration tests on specific data sources. If you add a test file for mssql, you should be able to use the existing `MSSQLDatasourceTestConfig` class following the pattern of other test files in this directory. If you need a hand, feel free to tag me or @wookasz

tyler-hoffman committed:
Merge branch 'develop' into fredwang1012

fredwang1012 committed:
change type from Any to str

fredwang1012 committed:
adding some commented out code

fredwang1012 committed:
gerge th 1.5.3remote-tracking branch 'great-expectations/develop' into fredwang1012

fredwang1012 committed:
Further Enhanced MSSQL regex

fredwang1012 committed:
further enhance regex to like

fredwang1012 committed:
Proper SQL Server Support Added

fredwang1012 committed:
Added tests for mssql

fredwang1012 committed:
Merge remote-tracking branch 'great-expectations/develop' into fredwang1012

fredwang1012 committed:
Remove unnecessary files from PR

fredwang1012 commented:
Hi @tyler-hoffman, I have updated this PR with your requested changes and also added some code to make SQL Server handling more consistent with how Snowflake is handled. Please let me know if I need to change anything else. Thanks.

tyler-hoffman commented on great_expectations/datasource/fluent/sources.py:
```diff
@@ -455,6 +453,39 @@ def _datasource_passed_in(
             f"positional argument: {name_or_datasource}, kwargs: {kwargs}"
         )
 
+    def _get_mssql_datasource_type(
```
Probably just a gap in my understanding: what's the need for this specific handling?

Requested change: calling `get_mssql_datasource_type` in general code seems off. Maybe there's something more appropriate to call this: get_specific_datasource_type or something along those lines?

tyler-hoffman commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -998,51 +1012,114 @@ class TableAsset(_SQLAsset):
-
-        # We reimport sqlalchemy from our compatability layer because we make
-        # quoted_name a top level import there.
+    def _resolve_quoted_name(cls, table_name: str) -> str:  # Returns str 
```
nit: no need to comment the return type when it's in the signature.

tyler-hoffman commented on great_expectations/datasource/fluent/sql_datasource.py:
```diff
@@ -1285,24 +1417,26 @@ def add_table_asset(
         name: str,
         table_name: str = "",
-        schema_name: Optional[str] = None,
+        schema_name: Optional[str] = MISSING,  # type: ignore[assignment] # sentinel value
```
Is this change needed for this PR? Why not just use None for this? If we do need the sentinal, let's add its type as a possible input here so we don't need to type ignore.

tyler-hoffman commented on great_expectations/datasource/fluent/sqlserver_datasource.py:
```diff
@@ -0,0 +1,99 @@
+
+
+@public_api
+class SQLServerDatasource(SQLDatasource):
```
We've called this MSSQL elsewhere. I'd recommend sticking with that naming for now. Maybe there's an argument to update globally to SQLServer down the road.

tyler-hoffman reviewed (commented):
Thanks for the updates! I added a couple new comments, and it looks like a few from last time I looked at this didn't get summited, so apologies on some of these comments coming in later than intended! Most of this looks solid, but a couple outstanding thoughts:
* I see a couple files deleted (including sqla execution engine) - was that unintentional? (this one is blocking)
* I think we should stick with the naming convention of mssql, rather than sql server, just for consistency. LMK if there's a nuance there that I should take into account.

fredwang1012 committed:
Merge branch 'develop' into fredwang1012

fredwang1012 committed:
Merge branch 'develop' into fredwang1012

wookasz commented:
Closing for lack of activity. Please don't hesitate to reopen if there are additional changes or if you feel this was done in error.
